### PR TITLE
add special handling for the route policy of the default VPC

### LINKED ISSUE #3102: 新增策略路由把之前的路由删除了
# Expected Behavior
新增一条策略路由
# Actual Behavior
之前的几条路由删除了
# Steps to Reproduce the Problem
1. 之前的路由
Routing Policies
     31000                           ip4.dst == 10.244.0.0/16           allow
     31000                           ip4.dst == 100.64.0.0/16           allow
     30000                           ip4.dst == 172.16.10.180         reroute                100.64.0.2
     30000                           ip4.dst == 172.16.10.189         reroute                100.64.0.3
     30000                            ip4.dst == 172.16.10.27         reroute                100.64.0.4
     29000                ip4.src == $ovn.default.master1_ip4         reroute                100.64.0.2
     29000                ip4.src == $ovn.default.master2_ip4         reroute                100.64.0.4
     29000                ip4.src == $ovn.default.master3_ip4         reroute                100.64.0.3

2.kubectl apply -f vpc_default_route.yam 添加路由
kind: Vpc
apiVersion: kubeovn.io/v1
metadata:
  name: ovn-cluster
spec:
  policyRoutes:
    - action: allow
      match: ip4.dst==10.0.0.0/16
      priority: 31000
3.之后的路由
Routing Policies
     31000                           ip4.dst == 10.244.0.0/16           allow
     31000                           ip4.dst == 100.64.0.0/16           allow
     31000                               ip4.dst==10.0.0.0/16           allow
     29000                ip4.src == $ovn.default.master1_ip4         reroute                100.64.0.2
     29000                ip4.src == $ovn.default.master2_ip4         reroute                100.64.0.4
     29000                ip4.src == $ovn.default.master3_ip4         reroute                100.64.0.3

# Additional Info
有一条路由也被删了
之前的路由
[root@master1 ~]# k ko nbctl lr-route-list ovn-cluster
IPv4 Routes
Route Table <main>:
              10.0.0.0/16               169.254.0.5 dst-ip
                0.0.0.0/0                100.64.0.1 dst-ip
之后的路由
IPv4 Routes
Route Table <main>:
                0.0.0.0/0                100.64.0.1 dst-ip
- Kubernetes version:
1.24.10
  **Output of `kubectl version`:**

  ```
  (paste your output here)
  ```

- kube-ovn version:
  1.11.8
  ```
  (paste your output here)
  ```

- operation-system/kernel version:
  "CentOS Linux 8"
5.15.63-1.el8.x86_64
  **Output of `awk -F '=' '/PRETTY_NAME/ { print $2 }' /etc/os-release`:**
  **Output of `uname -r`:**
  ```
  (paste your output here)
  ```

### PR DESCRIPTION
**Pull request by qiutingjun**
- [x] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

What type of this PR
Examples of user facing changes:

- Bug fixes

some policies are not maintained by codes in vpc.go and not reflected in the vpc spec.
while updating default vpc, these policies will be accidentally deleted.

system policy and static route synchronize with ovn.
custom policy and static route synchronize with spec.

### Which issue(s) this PR fixes:
Fixes #3102 

### <samp>🤖 Generated by Copilot at 21ec276</samp>

Refactor the code to use custom resource definitions (CRDs) to manage static and policy routes in virtual private clouds (VPCs). Update the controller logic and wrapper functions to create, update, and delete the CRD objects along with the OVN API calls. Modify the files `pkg/controller/init.go`, `pkg/controller/node.go`, `pkg/controller/ovn-ic.go`, `pkg/controller/pod.go`, `pkg/controller/subnet.go`, and `pkg/controller/gc.go`.

### <samp>🤖 Generated by Copilot at 21ec276</samp>

> _Sing, O Muse, of the mighty refactor that the coders wrought_
> _To use CRDs for routes in VPCs, a noble task they sought_
> _They wrapped the calls to `ovn-nbctl` and OVN-NB API_
> _With functions that create and delete the CRDs on the fly_

### <samp>🤖 Generated by Copilot at 21ec276</samp>

*  Refactor the code to use CRDs to manage static and policy routes in VPCs by replacing direct calls to OVN-NB API with wrapper functions that also create or delete the corresponding CRD objects ([link](https://github.com/kubeovn/kube-ovn/pull/3194/files?diff=unified&w=0#diff-efdb6355f2187b31e5e61610c24bf4d446a23d814d51448c170a91ef96ae1e89L651-R657), [link](https://github.com/kubeovn/kube-ovn/pull/3194/files?diff=unified&w=0#diff-92aa62435bc2b1b2be34f478f650597068bcf35b8d5975d119f91cf78ac54988L717-R747), [link](https://github.com/kubeovn/kube-ovn/pull/3194/files?diff=unified&w=0#diff-92aa62435bc2b1b2be34f478f650597068bcf35b8d5975d119f91cf78ac54988L739-R755), [link](https://github.com/kubeovn/kube-ovn/pull/3194/files?diff=unified&w=0#diff-8a92aa9e5847a415c4c90271fa845087d27a67fd272b8efc2636b06d52cb3619L269-R287), [link](https://github.com/kubeovn/kube-ovn/pull/3194/files?diff=unified&w=0#diff-8a92aa9e5847a415c4c90271fa845087d27a67fd272b8efc2636b06d52cb3619L1021-R1040), [link](https://github.com/kubeovn/kube-ovn/pull/3194/files?diff=unified&w=0#diff-8a92aa9e5847a415c4c90271fa845087d27a67fd272b8efc2636b06d52cb3619L1193-R1233), [link](https://github.com/kubeovn/kube-ovn/pull/3194/files?diff=unified&w=0#diff-173d2c52eb78053b84437284c077aba7452636bc799728cc475cc343ea70c2ceL375-R381), [link](https://github.com/kubeovn/kube-ovn/pull/3194/files?diff=unified&w=0#diff-173d2c52eb78053b84437284c077aba7452636bc799728cc475cc343ea70c2ceL487-R501), [link](https://github.com/kubeovn/kube-ovn/pull/3194/files?diff=unified&w=0#diff-173d2c52eb78053b84437284c077aba7452636bc799728cc475cc343ea70c2ceL494-R516), [link](https://github.com/kubeovn/kube-ovn/pull/3194/files?diff=unified&w=0#diff-449e8d1bd46b9c978e208dde912a1d7a76d17ce553067f8d0a67a27f17338d26L801-R808), [link](https://github.com/kubeovn/kube-ovn/pull/3194/files?diff=unified&w=0#diff-449e8d1bd46b9c978e208dde912a1d7a76d17ce553067f8d0a67a27f17338d26L852-R863), [link](https://github.com/kubeovn/kube-ovn/pull/3194/files?diff=unified&w=0#diff-449e8d1bd46b9c978e208dde912a1d7a76d17ce553067f8d0a67a27f17338d26L954-R970), [link](https://github.com/kubeovn/kube-ovn/pull/3194/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L1487-R1494), [link](https://github.com/kubeovn/kube-ovn/pull/3194/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L1819-R1829), [link](https://github.com/kubeovn/kube-ovn/pull/3194/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L2233-R2258), [link](https://github.com/kubeovn/kube-ovn/pull/3194/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L2271-R2302), [link](https://github.com/kubeovn/kube-ovn/pull/3194/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L2287-R2318), [link](https://github.com/kubeovn/kube-ovn/pull/3194/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L2327-R2358), [link](https://github.com/kubeovn/kube-ovn/pull/3194/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L2353-R2405), [link](https://github.com/kubeovn/kube-ovn/pull/3194/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L2380-R2423), [link](https://github.com/kubeovn/kube-ovn/pull/3194/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L2404-R2447), [link](https://github.com/kubeovn/kube-ovn/pull/3194/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L2534-R2604), [link](https://github.com/kubeovn/kube-ovn/pull/3194/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L2550-R2621), [link](https://github.com/kubeovn/kube-ovn/pull/3194/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L2655-R2724))
* Remove the unused import of `ovnnb` package in `node.go` as part of the code cleanup ([link](https://github.com/kubeovn/kube-ovn/pull/3194/files?diff=unified&w=0#diff-8a92aa9e5847a415c4c90271fa845087d27a67fd272b8efc2636b06d52cb3619L25))

### CODE REVIEW

File: pkg/controller/vpc.go [hunk 1]
```diff
@@ -269,157 +292,187 @@ func (c *Controller) handleAddOrUpdateVpc(key string) error {
...
-	if err != nil {
-		klog.Errorf("failed to get vpc %s static route list, %v", vpc.Name, err)
-		return err
-	}
-	rtbs := c.getRouteTablesByVpc(vpc)
-	targetRoutes := vpc.Spec.StaticRoutes
-	if vpc.Name == c.config.ClusterRouter {
-		if _, ok := rtbs[util.MainRouteTable]; !ok {
-			rtbs[util.MainRouteTable] = nil
+	{
```
Comment (bobz965) [inline pkg/controller/vpc.go:295]: 感觉多了一层缩进

File: pkg/controller/vpc.go [hunk 2]
```diff
@@ -269,157 +292,187 @@ func (c *Controller) handleAddOrUpdateVpc(key string) error {
...
+			gatewayV4, gatewayV6 := util.SplitStringIP(joinSubnet.Spec.Gateway)
+			if gatewayV4 != "" {
+				for talbe := range staticRouteMapping {
+					staticRouteTargets = append(
+						staticRouteTargets,
+						&kubeovnv1.StaticRoute{
+							Policy:     kubeovnv1.PolicyDst,
+							CIDR:       "0.0.0.0/0",
+							NextHopIP:  gatewayV4,
+							RouteTable: talbe,
```
Comment (bobz965) [inline pkg/controller/vpc.go:332]: talbe --> table

File: pkg/controller/vpc.go [hunk 3]
```diff
@@ -269,157 +292,187 @@ func (c *Controller) handleAddOrUpdateVpc(key string) error {
...
-			klog.Infof("vpc %s add static route: %+v", vpc.Name, item)
-			if err = c.OVNNbClient.AddLogicalRouterStaticRoute(
-				vpc.Name, item.RouteTable, convertPolicy(item.Policy), item.CIDR, nil, item.NextHopIP,
-			); err != nil {
-				klog.Errorf("failed to add normal static route to vpc %s , %v", vpc.Name, err)
+		for _, item := range routeNeedDel {
+			klog.Infof("vpc %s del static route: %v", vpc.Name, item)
+			policy := convertPolicy(item.Policy)
+			if err = c.OVNNbClient.DeleteLogicalRouterStaticRoute(vpc.Name, &item.RouteTable, &policy, item.CIDR, item.NextHopIP); err != nil {
+				klog.Errorf("del vpc %s static route failed, %v", vpc.Name, err)
```
Comment (bobz965) [inline pkg/controller/vpc.go:411]: 下面是 +v, 这里或许也应该

File: pkg/controller/vpc.go [hunk 4]
```diff
@@ -269,157 +292,187 @@ func (c *Controller) handleAddOrUpdateVpc(key string) error {
...
-		policyList, err := c.OVNNbClient.ListLogicalRouterPolicies(vpc.Name, -1, nil)
-		if err != nil {
-			klog.Errorf("failed to get vpc %s policy route list, %v", vpc.Name, err)
-			return err
-		}
-		policyRouteNeedDel, policyRouteNeedAdd, err := diffPolicyRoute(policyList, vpc.Spec.PolicyRoutes)
-		if err != nil {
-			klog.Errorf("failed to diff vpc %s policy route, %v", vpc.Name, err)
-			return err
+	// handle policy route
```
Comment (bobz965) [inline pkg/controller/vpc.go:437]: 多了一层缩进

File: pkg/controller/vpc.go [hunk 5]
```diff
@@ -526,33 +579,164 @@ func (c *Controller) handleAddOrUpdateVpc(key string) error {
 	return nil
 }
 
-func diffPolicyRoute(exist []*ovnnb.LogicalRouterPolicy, target []*kubeovnv1.PolicyRoute) (routeNeedDel, routeNeedAdd []*kubeovnv1.PolicyRoute, err error) {
-	existV1 := make([]*kubeovnv1.PolicyRoute, 0, len(exist))
-	for _, item := range exist {
-		existV1 = append(existV1, &kubeovnv1.PolicyRoute{
+func (c *Controller) addPolicyRouteToVpc(name string, policy *kubeovnv1.PolicyRoute, externalIDs map[string]string) error {
```
Comment (bobz965) [inline pkg/controller/vpc.go:582]: 感觉在命名上这里的 Vpc 替换为 DefaultVpc 更好些，也能更好的和自定义 Vpc的逻辑 对照，区分开

Comment (qiutingjun) [reply pkg/controller/vpc.go:582]: 这里并不是只是添加到默认VPC，只是把系统自动添加的动作收敛到了这里

**Review by bobz965 — APPROVED** [review]

**Comment by bobz965** [comment]
hi @zhangzujian @oilbeater could you help review?

File: pkg/controller/vpc.go [hunk 6]
```diff
@@ -653,20 +836,59 @@ func formatVpc(vpc *kubeovnv1.Vpc, c *Controller) error {
...
+	if list == nil {
+		return ""
+	}
+
+	var (
+		res []byte
+		err error
+	)
+
+	if res, err = json.Marshal(list); err != nil {
```
Comment (oilbeater) [inline pkg/controller/vpc.go:849]: need log the error

**Review by oilbeater — CHANGES_REQUESTED** [review]

**Review by oilbeater — APPROVED** [review]

**Pull request merged**